### PR TITLE
Package and deploy in one command

### PR DESCRIPTION
**Pull request by TheSriram**
- if the template is left unspecified, built artifacts are chosen and s3 bucket is optionally needed to deploy in one command.
- existing workflows to package and deploy do not break.

- [x] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/commands/deploy/command.py [hunk 1]
```diff
@@ -28,14 +31,14 @@
...
 
-@configuration_option(provider=TomlProvider(section="parameters"))
 @click.command(
     "deploy",
     short_help=SHORT_HELP,
     context_settings={"ignore_unknown_options": False, "allow_interspersed_args": True, "allow_extra_args": True},
     help=HELP_TEXT,
 )
-@template_click_option(include_build=False)
+@configuration_option(provider=TomlProvider(section="parameters"))
```
Comment (sanathkr) [inline samcli/commands/deploy/command.py]: is the order of decorators important?

File: samcli/commands/deploy/command.py [hunk 2]
```diff
@@ -28,14 +31,14 @@
...
+@template_click_option(include_build=True)
```
Comment (sanathkr) [inline samcli/commands/deploy/command.py]: Is this a breaking change or a new feature?

`sam build` -> `sam deploy` will now start to work, right?
`sam build` -> `sam package` -> `sam deploy without --template` will use the built template instead of the packaged template, right?

File: samcli/commands/deploy/command.py [hunk 3]
```diff
@@ -96,6 +99,14 @@
     "changes to be made to the stack. The default behavior is  to  return  a"
     "non-zero exit code.",
 )
+@click.option(
```
Comment (sanathkr) [inline samcli/commands/deploy/command.py]: I know we talked about this before, why is this explicitly required? can't we default to using JSON for the intermediate template?

File: samcli/commands/deploy/command.py [hunk 4]
```diff
@@ -154,27 +169,47 @@ def do_cli(
...
+            output_template_file=output_template_file.name,
+            kms_key_id=kms_key_id,
+            use_json=use_json,
+            force_upload=force_upload,
+            metadata=metadata,
+            on_deploy=True,
+            region=region,
+            profile=profile,
+        ) as package_context:
+            package_context.run()
```
Comment (sanathkr) [inline samcli/commands/deploy/command.py]: why are you packaging *always*? This would be a breaking change

File: samcli/commands/package/exceptions.py [hunk 5]
```diff
@@ -72,3 +72,12 @@ def __init__(self, template_file, ex):
         super(PackageFailedError, self).__init__(
             message=message_fmt.format(template_file=self.template_file, ex=self.ex)
         )
+
+
+class NoSuchBucketError(UserException):
+    def __init__(self, **kwargs):
+        self.kwargs = kwargs
+
+        message_fmt = "\n S3 Bucket does not exist."
```
Comment (sanathkr) [inline samcli/commands/package/exceptions.py]: Space at start of line - right after `\n`.

Why did you remove the use `aws s3 mb` text from the message?

File: samcli/commands/package/package_context.py [hunk 6]
```diff
@@ -67,6 +68,7 @@ def __init__(
         self.metadata = metadata
         self.region = region
         self.profile = profile
+        self.on_deploy = on_deploy
```
Comment (sanathkr) [inline samcli/commands/package/package_context.py]: so this is the special case for package happening as part of deploy?

File: samcli/commands/deploy/command.py (hunk 1 above)
Comment (TheSriram) [reply samcli/commands/deploy/command.py]: not after, we made template `is_eager=True`

File: samcli/commands/deploy/command.py (hunk 2 above)
Comment (TheSriram) [reply samcli/commands/deploy/command.py]: Yes. but today sam deploy requires --template-file to be specified.

being able to run sam deploy without --template-file is a new feature.

File: samcli/commands/deploy/command.py (hunk 3 above)
Comment (TheSriram) [reply samcli/commands/deploy/command.py]: Sure, we can. user had the option with package, do we gain anything by removing it from deploy?

File: samcli/commands/deploy/command.py (hunk 4 above)
Comment (TheSriram) [reply samcli/commands/deploy/command.py]: No. its an attempt to package, if the codeuri's already contain packaged contents, we do not repackage.

File: samcli/commands/package/exceptions.py (hunk 5 above)
Comment (TheSriram) [reply samcli/commands/package/exceptions.py]: `sam bootstrap` should come here logically.

File: samcli/commands/package/package_context.py (hunk 6 above)
Comment (TheSriram) [reply samcli/commands/package/package_context.py]: I'm just muting some output that comes on the terminal, afaik.

File: samcli/commands/deploy/command.py (hunk 1 above)
Comment (sanathkr) [reply samcli/commands/deploy/command.py]: okay

File: samcli/commands/deploy/command.py (hunk 2 above)
Comment (sanathkr) [reply samcli/commands/deploy/command.py]: Ah okay. We have to consolidate the myriad of options sometime in the future.

File: samcli/commands/deploy/command.py (hunk 4 above)
Comment (sanathkr) [reply samcli/commands/deploy/command.py]: This is skirting the lines of a breaking change. Is there any other mechanism to create two clear code paths, one for previous experience and other for new one?

And, this relies on the fact that `sam package` is idempotent. That's an undocumented assumption of `sam package`.

File: samcli/commands/package/exceptions.py (hunk 5 above)
Comment (sanathkr) [reply samcli/commands/package/exceptions.py]: right. can you add that?

File: samcli/commands/package/package_context.py (hunk 6 above)
Comment (sanathkr) [reply samcli/commands/package/package_context.py]: sure, but this feels hacky. May be in the future we will find more general use of this variable.

File: samcli/commands/deploy/command.py (hunk 3 above)
Comment (sanathkr) [reply samcli/commands/deploy/command.py]: There is no obvious user benefit of this flag in `sam deploy`. Why expose then?

File: samcli/commands/deploy/command.py (hunk 4 above)
Comment (TheSriram) [reply samcli/commands/deploy/command.py]: https://github.com/awslabs/aws-sam-cli/issues/1533. Added this question here.
